src/Faker/Provider/pl_PL/LicensePlate.php

```diff
@@ -0,0 +1,542 @@
+            'Służba Celno-Skarbowa' => ['CA', 'CB', 'CC', 'CD', 'CE', 'CF', 'CG', 'CH', 'CJ', 'CK', 'CL', 'CM', 'CN', 'CO', 'CP', 'CR'],
+            'Agencja Bezpieczeństwa Wewnętrznego' => ['K'],
+            'Agencja Wywiadu' => ['K'],
+            'Służba Kontrwywiadu Wojskowego' => ['M'],
+            'Służba Wywiadu Wojskowego' => ['M'],
+            'Policja' => ['PA', 'PB', 'PC', 'PD', 'PE', 'PF', 'PG', 'PH', 'PJ', 'PK', 'PL', 'PL', 'PL', 'PL', 'PL', 'PM', 'PN', 'PP', 'PS', 'PT', 'PU', 'PW', 'PZ'],
+            'Straż Graniczna' => ['WA', 'WK'],
+        ],
+    ];
+
+    /**
+     * @var array list of regex expressions matching Polish license plate suffixess when county code is 1 character long.
+     */
+    protected static $plateSuffixesGroup1 = [
+        '\d{5}',
+        '\d{4}[A-P,R-Z]',
+        '\d{3}[A-P,R-Z]{2}',
+        '[1-9][A-P,R-Z]\d{3}',
+        '[1-9][A-P,R-Z]{2}\d{2}',
+    ];
+
+    /**
+     * @var array list of regex expressions matching Polish license plate suffixess when county code is 2 characters long.
+     */
+    protected static $plateSuffixesGroup2 = [
+        '[A-P,R-Z]\d{3}',
+        '\d{2}[A-P,R-Z]{2}',
+        '[1-9][A-P,R-Z]\d{2}',
+        '\d{2}[A-P,R-Z][1-9]',
+        '[1-9][A-P,R-Z]{2}[1-9]',
+        '[A-P,R-Z]{2}\d{2}',
+        '\d{5}',
+        '\d{4}[A-P,R-Z]',
+        '\d{3}[A-P,R-Z]{2}',
+        '[A-P,R-Z]\d{2}[A-P,R-Z]',
+        '[A-P,R-Z][1-9][A-P,R-Z]{2}',
+    ];
+
+    /**
+     * Generates random license plate.
+     *
+     * @param  array|null $voivodeships
+     * @param  array|null $counties
+     *
+     * @return string
+     */
+    public static function licensePlate(?array $voivodeships = null, ?array $counties = null): string
+    {
+        $availableVoivodeships = array_keys(static::$voivodeships);
+        $voivodeships = array_filter($voivodeships ?? [], function ($value) use ($availableVoivodeships) {
+            return in_array($value, $availableVoivodeships);
+        });
+
+        if (count($voivodeships)) {
+            $voivodeship = static::randomElement($voivodeships);
+
+            $voivodeshipCode = static::$voivodeships[$voivodeship];
+        } else {
+            $voivodeshipCode = static::randomElement(static::$voivodeships);
+        }
+
+        $availableCounties = array_keys(static::$counties[$voivodeshipCode]);
+        $counties = array_filter($counties ?? [], function ($value) use ($availableCounties) {
+            return in_array($value, $availableCounties);
+        });
+
+        if (count($counties)) {
+            $countyCodes = static::$counties[$voivodeshipCode][static::randomElement($counties)];
+        } else {
+            $countyCodes = static::randomElement(static::$counties[$voivodeshipCode]);
+        }
+
+        $countyCode = static::randomElement($countyCodes);
+
+        if (strlen($countyCode) === 1) {
+            $suffix = static::regexify(static::randomElement(static::$plateSuffixesGroup1));
```

Comment:
Try to use ternary operator instead to avoid code duplication.

Reply:
I feel like this function should be split up into multiple functions.

Reply:
I've refactored a little bit.